pup-0003.md

```diff
@@ -0,0 +1,154 @@
+be clickable into a redmine query.
+* Use an associated commit. With this a redmine issue will show the original
+commits and any cherry-picks. We could have a new commit keyword so a commit
+could contain "cherrypick #1234" which would associate the cherry-pick commit
+with issue 1234. It would not modify the state of the issue so whatever state
+it's in, e.g. CLOSED - CURRENT RELEASE or MODIFIED would remain as-is.
+* Use a custom field to track the cherry-picks. This is convenient for release
+engineering queries, but would not show the actual cherry-pick commits
+themselves.
+
+
+## Drawbacks
+
+Git is really built for changes to be merged between branches. For example,
+when cherry-picking a change, the commit is assigned a new id but when merging
+the commit, it retains the same id. It makes looking up the commit to see what
+branches it's been applied to easier. If we cherry-pick, we have to track this
+information externally to git. This means also having to remember to update
+redmine.
+
+Switching to this model also creates some inconsistency. For example, if we
+adopt this for Pulp 3, then we still have to use merge forwarding for Pulp 2.
+Then we are stuck having to support two different strategies.
+
+Also, when creating bug fixes or hotfixes in the cherry-pick model, code will
+be developed against master. This could cause problems when the code is cherry
+picked to release branches such as the change relying on code that is in master
+but not in the release branch. This requires extra scrutiny and awareness on
+the part of the person cherry-picking a change back, to ensure that all code
+touched or used by a change exists on the release branch and behaves the same.
+Sometimes it will require cherry-picking intermediate changes that might not
+otherwise be obvious.
+
+Defaulting to master, and requiring extra steps to get a change onto the
+current release branch, is likely to result in fewer bug fixes making it onto
+current release branches. We can and should try to be diligent about
+cherry-picking fixes to the release branch, but the path of least resistance
+may prove too tempting at times.
+
+
+## Alternatives
+
+### Keeping the current model
+
+See the prior sections like "Motivation" and "Drawbacks" for a discussion on this.
+
+### Merge forward less often
+
+Perhaps keep the current model but merge less often. This would help to prevent
+accidents and keep the git history cleaner. There are some drawbacks however to
+this. One would be an increased chance of merge conflicts if we merge less
+often. However, we could also use Github's PR functionality to allow for code
+reviews of merges. This would also help to decrease accidental merges.
+
+
+## Unresolved Questions
+
+* If we adopt this proposal, how do we track in redmine what release branches a
+  fix has been applied to?
+* How do we transition to this model? Should we use it only for Pulp 3 for
+  instance?
```

Comment:
the idea of using 2 git workflows does not inspire me much. Previously we've mentioned that among drawbacks of current approach, community users do not know against which branch to open a pr. I with with adoption and usage of 2 workflows would blow their mind completely.